Flaky test: test_proxy_starts_and_routes_traffic — mitmdump 10s readiness timeout starves under heavy xdist parallelism

## Summary

`tests/infrastructure/test_http_replay_mitm.py::test_proxy_starts_and_routes_traffic` intermittently fails with a `mitmdump` startup-readiness timeout when the full suite runs under high `pytest-xdist` parallelism. It passes reliably in isolation, so this is a flaky (resource-starvation) failure, not a real defect in the proxy manager.

## Observed failure

During a pre-commit run (full fast suite, 32-way xdist — failure reported on `gw31`):

```
[gw31] win32 -- Python 3.13.2 ...\.venv\Scripts\python.exe
E   clm.infrastructure.http_replay_mitm.proxy_manager.MitmproxyError: mitmdump did not become ready within 10.0s:
src/clm/infrastructure/http_replay_mitm/proxy_manager.py:321
...
= 1 failed, 5972 passed, 3 skipped, 1 xfailed, 4 warnings in 141.70s =
```

Re-running the same test in isolation passes immediately:

```
tests/infrastructure/test_http_replay_mitm.py::test_proxy_starts_and_routes_traffic PASSED  [1 passed in 2.08s]
```

(Surfaced while landing #183 — unrelated change; the failure blocked the pre-commit commit until a retry cleared it.)

## Root cause hypothesis

`ProxyManager._wait_for_ready` polls the listen port until `mitmdump` accepts a TCP connection, bounded by a **hardcoded** module constant:

```python
# src/clm/infrastructure/http_replay_mitm/proxy_manager.py:39
_STARTUP_TIMEOUT_SECONDS = 10.0
```

Spawning an out-of-process `mitmdump` (Python interpreter + addon load + port bind) is genuinely slow, and 10s is not generous when ~30 xdist workers are simultaneously starting subprocesses/kernels and contending for CPU and I/O. This is the same class of resource-starvation flake the project has hit before with fixed timeouts under xdist (cf. worker-registration starvation under 32-way xdist). The lever here is the **readiness budget**, not the code path being waited on.

Two aggravating factors specific to this test:

1. The constant is module-level and **not injectable** — `ProxyManager.__init__` takes no `startup_timeout`, so a test can't widen it for the contended case.
2. The test carries **no marker** (`slow`/`integration`/`e2e`), so it runs inside the fast pre-commit suite at full parallelism — exactly the worst case for a subprocess-startup race.

## Candidate remedies (pick per maintainer preference)

- **Make the readiness budget generous and/or injectable.** Raise `_STARTUP_TIMEOUT_SECONDS` to something that survives contention (e.g. 30s — readiness is poll-based, so a higher ceiling costs nothing on the happy path), and/or thread a `startup_timeout` through `ProxyManager.__init__` so callers/tests can tune it. Optionally honor an env override (`CLM_MITM_STARTUP_TIMEOUT`).
- **Reduce contention for this test.** Mark it so it doesn't run at full fast-suite parallelism (e.g. an `xdist_group`/serialization marker, or move it behind an `integration`/`slow` marker so the pre-commit fast suite skips it — it is already a documented smoke test for the parked mitmproxy prototype).

Widening the wait (not bumping a tight ceiling repeatedly) is the durable fix: the thing being polled — `mitmdump` process startup — legitimately takes longer under load.

## Repro notes

- Environment: Windows, Python 3.13, `pytest-xdist -n auto` (32 workers on this machine).
- Non-deterministic: requires enough concurrent load to delay `mitmdump` startup past 10s. Easiest trigger is the full fast suite under high `-n`.
- Not reproducible in isolation or at low parallelism.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

--- a/docs/claude/issue-165-p5-readiness.md
+++ b/docs/claude/issue-165-p5-readiness.md
@@ -186,6 +186,12 @@ $env:CLM_HTTP_REPLAY_TRANSPORT = "mitmproxy"
 $Repro = "C:\Users\tc\Programming\Python\Tests\clm-bug-repros\issue-143-cassette-connection-pool-deadlock"
 ```
 
+Optional tuning: `CLM_MITM_STARTUP_TIMEOUT` (seconds) overrides the proxy
+readiness budget. The default is 30s; raise it only if `mitmdump did not become
+ready` fires on a loaded host. The timeout error names this var, and
+distinguishes an overloaded-but-still-starting proxy from a genuine crash
+(issue #184).
+
 | Gate | How to run | PASS criterion |
 |---|---|---|
 | 3 byte-identity, 4 tooling | `pytest tests/infrastructure/test_http_replay_mitm_cassette_format.py` | green (skip-safe w/o mitmdump) |

--- a/src/clm/infrastructure/http_replay_mitm/proxy_manager.py
+++ b/src/clm/infrastructure/http_replay_mitm/proxy_manager.py
@@ -34,9 +34,48 @@ logger = logging.getLogger(__name__)
 # Path to the addon module so we can pass it to ``mitmdump --scripts``.
 _ADDON_PATH = Path(__file__).parent / "addon.py"
 
-# How long we wait for the proxy to accept TCP connections before
-# giving up. Local startup is typically <500ms; cap conservatively.
-_STARTUP_TIMEOUT_SECONDS = 10.0
+# How long we wait for the proxy to accept TCP connections before giving up.
+# Local mitmdump startup is typically <500ms, but a loaded host — heavy xdist
+# parallelism, or background OS work like Windows installing updates — can push
+# the port-bind well past that. The readiness poll returns the instant the port
+# accepts, so a generous default is nearly free on the happy path; the full
+# budget only elapses for an *alive-but-not-yet-listening* process (a genuine
+# crash short-circuits via ``poll()``). ``CLM_MITM_STARTUP_TIMEOUT`` (seconds)
+# overrides it for CI or build hosts that need more headroom.
+_DEFAULT_STARTUP_TIMEOUT_SECONDS = 30.0
+_STARTUP_TIMEOUT_ENV = "CLM_MITM_STARTUP_TIMEOUT"
+
+
+def _startup_timeout_seconds() -> float:
+    """Resolve the proxy-readiness budget, honouring ``CLM_MITM_STARTUP_TIMEOUT``.
+
+    Returns the env override when it parses to a positive number, otherwise
+    ``_DEFAULT_STARTUP_TIMEOUT_SECONDS``. An unparseable or non-positive value
+    is ignored with a warning so a typo can't silently disable the wait.
+    """
+    raw = os.environ.get(_STARTUP_TIMEOUT_ENV)
+    if raw is None:
+        return _DEFAULT_STARTUP_TIMEOUT_SECONDS
+    try:
+        value = float(raw)
+    except ValueError:
+        logger.warning(
+            "Ignoring non-numeric %s=%r; using default %.1fs.",
+            _STARTUP_TIMEOUT_ENV,
+            raw,
+            _DEFAULT_STARTUP_TIMEOUT_SECONDS,
+        )
+        return _DEFAULT_STARTUP_TIMEOUT_SECONDS
+    if value <= 0:
+        logger.warning(
+            "Ignoring non-positive %s=%r; using default %.1fs.",
+            _STARTUP_TIMEOUT_ENV,
+            raw,
+            _DEFAULT_STARTUP_TIMEOUT_SECONDS,
+        )
+        return _DEFAULT_STARTUP_TIMEOUT_SECONDS
+    return value
+
 
 # How long we wait for graceful shutdown before sending SIGKILL/terminate.
 # mitmdump flushes flow streams on exit, so we want enough time for that.
@@ -305,7 +344,8 @@ class MitmproxyManager:
         assert self._process is not None
         assert self.listen_port is not None
 
-        deadline = time.monotonic() + _STARTUP_TIMEOUT_SECONDS
+        timeout = _startup_timeout_seconds()
+        deadline = time.monotonic() + timeout
         while time.monotonic() < deadline:
             if self._process.poll() is not None:
                 output = self._drain_output()
@@ -318,8 +358,19 @@ class MitmproxyManager:
             except OSError:
                 time.sleep(0.05)
         output = self._drain_output()
+        # Distinguish the two ways the budget can elapse so the error is
+        # actionable. A still-running process bound too slowly — the host is
+        # likely overloaded and the remedy is more time, not a code change. A
+        # process that has since exited is a genuine startup failure.
+        if self._process.poll() is None:
+            raise MitmproxyError(
+                f"mitmdump did not become ready within {timeout:.1f}s and is still "
+                f"starting — the host may be overloaded. Set {_STARTUP_TIMEOUT_ENV} "
+                f"(seconds) to allow more time if this recurs.\n{output}"
+            )
         raise MitmproxyError(
-            f"mitmdump did not become ready within {_STARTUP_TIMEOUT_SECONDS:.1f}s:\n{output}"
+            f"mitmdump exited (rc={self._process.returncode}) without becoming ready "
+            f"within {timeout:.1f}s:\n{output}"
         )
 
     def _drain_output(self) -> str:

--- a/tests/infrastructure/test_http_replay_mitm_manager.py
+++ b/tests/infrastructure/test_http_replay_mitm_manager.py
@@ -12,6 +12,8 @@ import subprocess
 import sys
 import time
 
+import pytest
+
 from clm.infrastructure.http_replay_mitm import proxy_manager
 from clm.infrastructure.http_replay_mitm.proxy_manager import MitmproxyManager
 
@@ -143,3 +145,90 @@ class TestStartCommandTraceDir:
     def test_no_trace_dir_omits_clm_trace_dir_set(self, monkeypatch) -> None:
         cmd = self._captured_cmd(monkeypatch)
         assert not any("clm_trace_dir=" in str(part) for part in cmd)
+
+
+def _raise_oserror(*_args, **_kwargs):
+    """Stand-in for ``socket.create_connection`` that always refuses."""
+    raise OSError("connection refused")
+
+
+class _ReadyFakeProc:
+    """Minimal ``Popen`` stand-in for ``_wait_for_ready`` diagnostics.
+
+    ``rc=None`` models an alive-but-not-listening process (the overload case);
+    a non-None ``rc`` models a process that has already exited.
+    """
+
+    def __init__(self, rc: int | None = None) -> None:
+        self._rc = rc
+        self.returncode = rc
+
+    def poll(self) -> int | None:
+        return self._rc
+
+
+class TestStartupTimeoutResolution:
+    """``_startup_timeout_seconds`` honours ``CLM_MITM_STARTUP_TIMEOUT`` (issue #184).
+
+    The readiness budget is generous by default because a loaded host can bind
+    the port well past the old 10s, but it stays overridable for hosts that
+    need even more headroom. A typo must not silently disable the wait.
+    """
+
+    def test_default_when_unset(self, monkeypatch) -> None:
+        monkeypatch.delenv(proxy_manager._STARTUP_TIMEOUT_ENV, raising=False)
+        assert (
+            proxy_manager._startup_timeout_seconds()
+            == proxy_manager._DEFAULT_STARTUP_TIMEOUT_SECONDS
+        )
+
+    def test_valid_override_is_used(self, monkeypatch) -> None:
+        monkeypatch.setenv(proxy_manager._STARTUP_TIMEOUT_ENV, "45")
+        assert proxy_manager._startup_timeout_seconds() == 45.0
+
+    def test_non_numeric_falls_back_to_default(self, monkeypatch) -> None:
+        monkeypatch.setenv(proxy_manager._STARTUP_TIMEOUT_ENV, "soon")
+        assert (
+            proxy_manager._startup_timeout_seconds()
+            == proxy_manager._DEFAULT_STARTUP_TIMEOUT_SECONDS
+        )
+
+    def test_non_positive_falls_back_to_default(self, monkeypatch) -> None:
+        monkeypatch.setenv(proxy_manager._STARTUP_TIMEOUT_ENV, "0")
+        assert (
+            proxy_manager._startup_timeout_seconds()
+            == proxy_manager._DEFAULT_STARTUP_TIMEOUT_SECONDS
+        )
+
+
+class TestWaitForReadyDiagnostics:
+    """The readiness timeout message distinguishes overload from a crash (issue #184)."""
+
+    def test_overloaded_process_still_starting_points_to_env_knob(self, monkeypatch) -> None:
+        # Tiny budget + a port that never accepts exercises the timeout path
+        # without waiting the real default.
+        monkeypatch.setenv(proxy_manager._STARTUP_TIMEOUT_ENV, "0.2")
+        monkeypatch.setattr(proxy_manager.socket, "create_connection", _raise_oserror)
+        mgr = _manager()
+        mgr._process = _ReadyFakeProc(rc=None)  # alive, never binds
+        mgr.listen_port = 65000
+
+        with pytest.raises(proxy_manager.MitmproxyError) as excinfo:
+            mgr._wait_for_ready()
+        msg = str(excinfo.value)
+        assert "overloaded" in msg.lower()
+        assert proxy_manager._STARTUP_TIMEOUT_ENV in msg
+        assert "0.2s" in msg  # reports the resolved budget, not the default
+
+    def test_exited_process_reports_crash_not_overload(self, monkeypatch) -> None:
+        monkeypatch.setattr(proxy_manager.socket, "create_connection", _raise_oserror)
+        mgr = _manager()
+        mgr._process = _ReadyFakeProc(rc=7)  # already exited
+        mgr.listen_port = 65000
+
+        with pytest.raises(proxy_manager.MitmproxyError) as excinfo:
+            mgr._wait_for_ready()
+        msg = str(excinfo.value)
+        assert "exited during startup" in msg
+        assert "rc=7" in msg
+        assert "overloaded" not in msg.lower()

